internal/oidc/keys.go

```diff
@@ -0,0 +1,195 @@
+package oidc
+
+import (
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/ory/fosite/token/jwt"
+	"gopkg.in/square/go-jose.v2"
+
+	"github.com/authelia/authelia/internal/configuration/schema"
+	"github.com/authelia/authelia/internal/utils"
+)
+
+// NewKeyManagerWithConfiguration when provided a schema.OpenIDConnectConfiguration creates a new KeyManager and adds an
+// initial key to the manager.
+func NewKeyManagerWithConfiguration(configuration *schema.OpenIDConnectConfiguration) (manager *KeyManager, err error) {
+	manager = NewKeyManager()
+
+	_, _, err = manager.AddActiveKeyData(configuration.IssuerPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return manager, nil
+}
+
+// NewKeyManager creates a new empty KeyManager.
+func NewKeyManager() (manager *KeyManager) {
+	manager = new(KeyManager)
+	manager.keys = map[string]*rsa.PrivateKey{}
+	manager.keySet = new(jose.JSONWebKeySet)
+
+	return manager
+}
+
+// Strategy returns the RS256JWTStrategy.
+func (m KeyManager) Strategy() (strategy *RS256JWTStrategy) {
+	return m.strategy
+}
+
+// GetKeySet returns the joseJSONWebKeySet containing the rsa.PublicKey types.
+func (m KeyManager) GetKeySet() (keySet *jose.JSONWebKeySet) {
+	return m.keySet
+}
+
+// GetActiveWebKey obtains the currently active jose.JSONWebKey.
+func (m KeyManager) GetActiveWebKey() (webKey *jose.JSONWebKey, err error) {
+	webKeys := m.keySet.Key(m.activeKeyID)
+	if len(webKeys) == 1 {
+		return &webKeys[0], nil
+	}
+
+	if len(webKeys) == 0 {
+		return nil, errors.New("could not find a key with the active key id")
+	}
+
+	return &webKeys[0], errors.New("multiple keys with the same key id")
+}
+
+// GetActiveKeyID returns the key id of the currently active key.
+func (m KeyManager) GetActiveKeyID() (keyID string) {
+	return m.activeKeyID
+}
+
+// GetActiveKey returns the rsa.PublicKey of the currently active key.
+func (m KeyManager) GetActiveKey() (key *rsa.PublicKey, err error) {
+	if key, ok := m.keys[m.activeKeyID]; ok {
+		return &key.PublicKey, nil
+	}
+
+	return nil, errors.New("failed to retrieve active public key")
+}
+
+// GetActivePrivateKey returns the rsa.PrivateKey of the currently active key.
+func (m KeyManager) GetActivePrivateKey() (key *rsa.PrivateKey, err error) {
+	if key, ok := m.keys[m.activeKeyID]; ok {
+		return key, nil
+	}
+
+	return nil, errors.New("failed to retrieve active private key")
+}
+
+// AddActiveKeyData adds a rsa.PublicKey given the key in the PEM string format, then sets it to the active key.
+func (m *KeyManager) AddActiveKeyData(data string) (key *rsa.PrivateKey, webKey *jose.JSONWebKey, err error) {
```

Comment:
I would rename this method to AddActivePrivateKeyData to make it clear that the data should be a private key.